### ISSUE TITLE
The threading model of webrtc-pc: When are effects of in-parallel stuff surfaced?

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-pc#2502]
JavaScript "pretending" to be single-threaded, things that happen either just happen (in the currently executing task), or they happen "in-parallel" (i.e. off the thread), and then the results are surfaced sometime in the future in a queued task.

According to my understanding of things, exposing the effects of something from an in-parallel operation before the queued task has executed would be a bug. I'm going to call this **_the state surfacing theorem_**. Feel free to challenge the validity of this theorem.

For example, `while (true) { console.log(sender.getSynchronizationSources()); }` should ALWAYS print the same result, because the thread is blocked and no queued task is able to execute to inform the JS thread that a new RTP packet has arrived.

Similarly, I would expect the following to print 0 forever and ever, because we are not letting the in-parallel SRD operation perform its queued task to create the JavaScript layer transceiver objects corresponding to the SDP m= sections:
```
pc.setRemoteDescription(offerThatCreatesATransceiver);
while (true) {
  console.log(pc.getTransceivers().length);
}
```

Why does this matter? Well JSEP has this to say about addTrack:

> If the PeerConnection is in the "have-remote-offer" state, the track will be attached to the first compatible transceiver that was created by the most recent call to setRemoteDescription() and does not have a local track.

But setRemoteDescription() executes in-parallel. And addTrack() executes immediately. So if I do `setRemoteDescription(); addTrack();`, has SRD happened when addTrack happens or not?

Based on recent discussions, I believe Chrome and Firefox has observably different implementations. I believe both of them violate _the state surfacing theorem_ but in different ways. Correct me if I'm wrong on Firefox's implementation.

- **Behavior A** (= what Chrome does): In-parallel operations previously queued happen first and synchronous operations happen second due to PostTasks ensuring a FIFO order. If you call SRD followed by addTrack, the JS thread is blocked until addTrack is performed, which only happens after the SRD operation has executed. The theorem is violated because what the result of the addTrack operation is could expose how SRD was executed.
- **Behavior B** (= probably what Firefox does?): In-parallel operations and JavaScript share the same set of transceiver objects, accessed by grabbing locks. The result of SRD + addTrack is undefined, because which operation modifies the set of transceiver objects first is racy. The theorem is violated in the cases when the race favors SRD happening first.
- **Behavior C** (= not implemented by any browser?): No races and synchronous operations truly happening synchronously - the effect of SRD is truly not visible until the posted task. That is, if SRD is immediately followed by addTrack, the addTrack operation MUST act as if SRD has not happened yet.

What is the correct behavior? Discuss.

**Comment by henbos** [comment]
My position:
- I am against Behavior A because it blocks the JS thread on "in-parallel" work, effectively making them not "in-parallel". Chrome's WebRTC implementation is violating Chromium design principles.
- I am against Behavior B because it is racy.
- I believe Behavior C is the "correct behavior" but I'm afraid fixing the implementations are not worth the benefits. Is there a compromise?

I am happy with updating the spec such that minimum work is required on behalf of browser vendors (ideally none). But right now, I feel like the correct behavior is either undefined or not what current implementations are doing today.

**Comment by henbos** [comment]
The implementation of **Behavior C** would require that we interpret the words "_the most recent call to setRemoteDescription()_" to mean "the most recent setRemoteDescription() operation that has completed and whose states have been surfaced to the JS thread".

**Comment by henbos** [comment]
Thoughts @jan-ivar ?

**Comment by jan-ivar** [comment]
As I cover in https://github.com/w3c/webrtc-pc/issues/2476, there are **_two_** places we special-case addTrack to _minimize # of transceivers_:
 1. In parallel SRD ([JSEP section 5.10](https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-25#section-5.10)): *"If ... there are RtpTransceivers ...  added ... by addTrack and are not associated ... and are not stopped, find the first ... Associate the found ... with the m= section ..."*
 2. In addTrack itself.

The first covers covers `pc.addTrack(track); pc.setRemoteDescription(offer)`
The second covers covers `await pc.setRemoteDescription(offer); pc.addTrack(track)`
So what covers `pc.setRemoteDescription(offer); pc.addTrack(track)`? TL;DR: neither I think.

The answer depends on the answer to **_two_** questions:
 1. Does JSEP's in parallel SRD read webrtc-pc's [set of transceivers](https://w3c.github.io/webrtc-pc/#transceivers-set)? That'd be a data race!
 2. Which addTrack algorithm are we running?

The answer to the first one is unclear. JSEP (oblivious to threads) is run in parallel by webrtc-pc, so this is a bug in webrtc-pc I claim.

The answer to the second one I think is covered:

> JSEP has this to say about addTrack:

That's a _concept method_ under [section 4](https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-24#section-4) which states: *"The actual API exposed in the W3C API may have somewhat different syntax, but SHOULD map easily to these concepts."*

Thus webrtc-pc's [addTrack](https://w3c.github.io/webrtc-pc/#dom-rtcpeerconnection-addtrack) is the algorithm that counts. It says: *"The steps below describe how to determine if an existing sender can be reused."* & describes that process entirely using the [set of transceivers](https://w3c.github.io/webrtc-pc/#transceivers-set), which is only [modified](https://w3c.github.io/webrtc-pc/#ref-for-transceivers-set-2) in SRD's queued task.

So the answer is:

> if I do setRemoteDescription(); addTrack();, has SRD happened when addTrack happens or not?

SRD says _"In parallel, start the process"_, so the JSEP process may have _started_, however, that does not matter to addTrack, which only looks at the [set of transceivers](https://w3c.github.io/webrtc-pc/#transceivers-set), which is only [modified](https://w3c.github.io/webrtc-pc/#ref-for-transceivers-set-2) in SRD's queued task (which hasn't happened yet).

But unless we fix the data race in JSEP's in parallel SRD, it may still pick it up, depending on how far along we are in that in-parallel process. 🤢

Ironically, as I mention at the bottom of https://github.com/w3c/webrtc-pc/issues/2476, if we fix the data race, then we have a tiny window where addTrack will not associate with available transceivers. Since it would have associated itself both a few milliseconds earlier _**and**_ a few milliseconds later, that seems dead wrong.

**Comment by henbos** [comment]
JSEP's in-parallel steps can't read webrtc-pc's set of transceivers, because that is a race. We agree.
Likewise, webrtc-pc can't read JSEP's set of transceivers, because that would also be a race. And also because this set is maintained by webrtc-pc, not by JSEP, and as such is only modified in JS tasks.

If JSEP's set of transceivers is not the same as webrtc-pc's, and webrtc-pc's is not the same as JSEP, that gives us our first premise:

**1. webrtc-pc and JSEP have separate sets of transceivers**

> It says: _"The steps below describe how to determine if an existing sender can be reused."_ & describes that process entirely using the [set of transceivers](https://w3c.github.io/webrtc-pc/#transceivers-set), which is only [modified](https://w3c.github.io/webrtc-pc/#ref-for-transceivers-set-2) in SRD's queued task.

We agree. This is both explicit by the spec today and is supported by premise 1).

> But unless we fix the data race in JSEP's in parallel SRD, it may still pick it up, depending on how far along we are in that in-parallel process.

I don't think so. While we know the sets of transceivers are different, and as such it does not matter if webrtc-pc's addTrack() has modified the webrtc-pc's set of transcievers yet or not, it is a valid question to ask if the JSEP's AddTrack() method has been executed yet.

I would argue that JSEP is a single-threaded toolkit used by webrtc-pc. It might be backed up by other threads doing the work like network and media, but the APIs of JSEP: AddTrack, SetRemoteDescription, etc, are written as if they were single-threaded.

Would you accept the following premise?

**2. The API of JSEP (its set of transceivers and which signaling state it is in) is a single-threaded toolkit**

If 2) is true we can do SetRemoteDescription and we can do AddTrack, but we cannot possibly do both at the same time. The only resonable thing to do is to queue performing AddTrack until after SRD is finished on the JSEP thread.

This gets interesting. webrtc-pc determines that a transceiver cannot be recycled because in webrtc-pc SRD has not happened yet. But JSEP, having done SRD first and AddTrack second, would necessarily conclude that there IS a transceiver to be recycled.

There are two ways to resolve the contradiction:
1. Make webrtc-pc's decision about which transceiver to recycle override JSEP's decision about which transceiver to recycle.
2. Make JSEP decide, but this would necessarily have to make the addTrack() method promise-based.

**Comment by henbos** [comment]
1) might lead to weirdness like attaching to a transceiver that is being stopped in parallel through SDP.
But I think it's too late to change the API to accommodate for 2).

**Comment by jan-ivar** [comment]
I want to address those good ideas, but first I'd like to separate interpretation from remedy, because I don't accept those premises as interpretations of the current spec.

**My interpretation is in https://github.com/w3c/webrtc-pc/issues/2476:** there's one set of transceivers in webrtc-pc that JSEP reads and modifies from a different thread.

> JSEP's in-parallel steps can't read webrtc-pc's set of transceivers, because that is a race.

Sure it can. Yes it's a race, and I agree we _probably_ need to fix it somehow, but it's what the spec says to do at the moment, because JSEP says _"RtpTransceivers of ... that were added to the PeerConnection by addTrack"_, and webrtc-pc doesn't intervene here.

I want to establish that A) Firefox is spec compliant here, and 2) the spec needs to be fixed. Because at least one Firefox engineer who worked hard to follow the spec, will have a lot of work to do to fix this with whatever we come up with, so I think we should be clear the problem is in the spec, not the interpretation.

Oh, and C) the intent has always been for this addTrack to get associated one way or another:
```js
pc.setRemoteDescription(offer); // note: no await
pc.addTrack(track, stream); // Supposed to always get associated
```
...even though even in the current spec with its direct cross-thread data writing, one _tiny window_ remains: An (e.g. onclick) addTrack called in a task queued immediately ahead of SRD's queued task for success, won't get associated.

**Comment by jan-ivar** [comment]
> **1. webrtc-pc and JSEP have separate sets of transceivers**

I agree with this as a remedy. We must avoid updating JS observable stuff off main-thread. But it's going to widen the "tiny window", so we need to eliminate that window somehow.

> 2. The API of JSEP (its set of transceivers and which signaling state it is in) is a single-threaded toolkit

I'm not sure it matters how we characterize JSEP here, as we cannot "queue" or chain _addTrack_, because it's a synchronous method. 

What I agree with is JSEP's intent here seems to clearly be C in [my previous comment](https://github.com/w3c/webrtc-pc/issues/2502#issuecomment-606668087).

I think we need to do (1) above, and then monkey patch C in webrtc-pc. It's going to get ugly.

**Comment by henbos** [comment]
> I want to address those good ideas, but first I'd like to separate interpretation from remedy, because I don't accept those premises as interpretations of the current spec.

That's OK. I suppose multiple interpretations are valid when we are discussing something that the authors of either spec probably never thought of. Let's see if we can find agreement on what we _want_ the spec(s) to say :)

> **My interpretation is in #2476:** there's one set of transceivers in webrtc-pc that JSEP reads and modifies from a different thread.

That's an understandable interpretation, but one that that implies races in both webrtc-pc and JSEP. I fear this interpretation will lead to a lot of monkey patching.

### **In response to what is possible (whether or not we think its a good idea)**

> > **2. The API of JSEP (its set of transceivers and which signaling state it is in) is a single-threaded toolkit**
> 
> I'm not sure it matters how we characterize JSEP here, as we cannot "queue" or chain addTrack, because it's a synchronous method.

Prepare for cans of worms, but... :) perhaps we can? Grab your favourite tinfoil hat and follow me!

Feel free to disagree with this interpretation or approach, but try to help me understand if this actually works... Chrome is doing something similar today except that it blocks the JS thread on the JSEP thread any time it tries to do something "synchronous" (which we obviously shouldn't do). That might avoid some of the complexity of the proposed interpretation which relies on "dummy objects" (we don't have them today):

You see, addTrack() cannot actually fail, and neither can any of the _synchronous_ control surfaces exposed by the [RTP Media API](https://w3c.github.io/webrtc-pc/#rtp-media-api). So you could return a dummy transceiver object and start the process of performing AddTrack() on the JSEP thread in-parallel, unblocking the JS thread which has obtained either an existing transceiver or a dummy transceiver for the to-be-created transceiver.

You want to do `transceiver.direction = 'inactive'` on the returned object, even though the JSEP corresponding object might not exist yet? That's fine, we just queue a task to update the corresponding JSEP transceiver's direction to `'inactive'`. In the meantime, the `[[Direction]]` internal slot will return `'inactive'` and there is no way for you to observe what the JSEP transceiver's direction is synchronously.

If you set the direction followed by createOffer(), you will see `inactive` reflected in the SDP because by the time CreateOffer() happens in the JSEP thread, JSEP will already have performed the direction setting. I imagine all communication JS -&gt; JSEP happens with PostTask, and all communication JSEP -&gt; JS happens with PostTask as well.

JS modifies things like "direction". JSEP modifies things like "currentDirection". It's OK to be out-of-sync for a few milliseconds, because we're not depending on the same states at the same time.

I'm not aware of any scenario where the fact that JS and JSEP transceivers are out-of-sync causes a problem except for when there is a synchronous method that relies on the result of asynchronous operations. I.e. SRD+addTrack. Is there any other example than addTrack where we have a synchronous operation depending on asynchronous events?

There are other synchronous methods, like `pc.close()` or `transceiver.stop()`. But these are `void`, and the effect on any JS-exposed attributes as a result of these methods are predictable and could be replaced by a "dummy object" if we want to avoid blocking the JS thread on the JSEP thread. Today, Chrome blocks the JS thread on the JSEP thread, but I imagine setting internal slots and dummies could allow us to stop blocking it.

### **In response to use case C)**

> What I agree with is JSEP's intent here seems to clearly be C in [my previous comment](https://github.com/w3c/webrtc-pc/issues/2502#issuecomment-606668087).

But JSEP also does not seem to be multi-thread aware, so in JSEP, any example of `SRD(); AddTrack();` would be the JS-equivalent of `await SRD(); addTrack();`. And for that, the problem we are discussing does not arise. I believe the problems we are discussing are original to webrtc-pc, because it is webrtc-pc that introduced multiple threads. Thus, it is webrtc-pc that has to ensure that we do not get any contradictions.

> ```
> pc.setRemoteDescription(offer); // note: no await
> pc.addTrack(track, stream); // Supposed to always get associated
> ```

In that example, if we want addTrack() to be associated with a transceiver that was created by the in-parallel operation SRD(), why are we not awaiting for SRD() before we do addTrack()?

Is this really a use case we need to support or is this an example of racy application logic? In which case, does it matter that it doesn't get associated?

The example of a user clicking a button after SRD but before the promise is resolved is even more random. The user clicking a button is inherently racy with other application code.

**Comment by henbos** [comment]
Walls of text aside, I would prefer if we could find a solution that doesn't require either browser to rewrite significant parts of their code. (Cost vs benefit)

It matters more to me that getTransceivers(), signalingState or other attributes do not get modified inside the same JS task than whether or not a transceiver gets associated if you do addTrack() while SDP is being processed in the background.

**Comment by jan-ivar** [comment]
Sorry I'm not familiar enough with how Chrome works to follow, but we can't break the synchronous nature of addTrack.

> why are we not awaiting for SRD() before we do addTrack()?

As definers of the API, it's our job to define what happens, not ask why it's done. ;)

But since you ask me to defend taking action: The reason I think we must solve this is this one:
```js
pc.setRemoteDescription(offer);
button.onclick = () => pc.addTrack(track, stream); // Must always get associated
```
 1. If addTrack happens right before SRD, then SRD associates it.
 2. If addTrack happens right after SRD success, then addTrack associates it.
 3. If addTrack happens right in the middle of SRD, then it is not associated (theoretically)

Since both 1 and 2 associate, it seems pretty darn obvious that 3 should as well. Especially since add

Note I haven't actually been able to _observe_ 3 in any implementation, as it's hard to test.
I think it might suffice to say something like _"If addTrack was called during the process to apply description, then start the process over."_ I'll do a PR.

**Comment by jan-ivar** [comment]
I'll try to write a PR for https://github.com/w3c/webrtc-pc/issues/2476 to copy the set of transceivers and mid.

**Comment by henbos** [comment]
> Since both 1 and 2 associate, it seems pretty darn obvious that 3 should as well.

Ack! I was mistakingly thinking that association didn't happen in that case, my bad.
Then we are in agreement. If association happens both before and after, then it must also associate if it happens during as well, anything else would be a foot-gun. Concluding that association must happen regardless, I will no longer accuse your hypothetical application of being racy - it should work.

> we can't break the synchronous nature of addTrack.

I still suspect that we _can_, but it gets complicated, and that implementation is theoretical at this point. If we can avoid it, I'm all for it!

**Comment by henbos** [comment]
In your view, if the following happens...

```
pc.setRemoteDescription(offer);  // no await
pc.addTrack(track);
```

Will addTrack() create a new transceiver that gets associated as part of SRD surfacing state changes?
Or will addTrack() grab an in-parallel created transceiver and associate it before JS knows SRD was done?

**Comment by jan-ivar** [comment]
The former, with #2504

**Comment by henbos** [comment]
I think the former and #2504 makes sense, but Chrome currently does the latter.
Assuming Chrome gets updated to [not expose the mid value prematurely](https://github.com/w3c/webrtc-pc/pull/2504#issuecomment-606874108), can you foresee any observable difference between doing the former or doing the latter?

If the end-result is the same, that we have an associated transceiver, then Chrome doing things sub-optimally would be a performance issue, not a spec-compliance issue. I'm hoping that to be the case.

**Comment by jan-ivar** [comment]
> Assuming Chrome gets updated to not expose the mid value prematurely, can you foresee any observable difference between doing the former or doing the latter?

No.
